BufferUtils: Optimize upload_untouched_skip_restart with AVX-512 paths

Whatcookie opened the pull request:
Uses vpcompress to vectorize this otherwise unvectorizeable loop
the u16 path needs AVX-512-ICL because vpcompressw isn't included in skylake-x level AVX-512
the u32 path is untested as I couldn't find any games that hit it

We use vcompress register to register, rather than directly to memory since there's a bug with vcompress to memory on zen4, which makes it exceedingly slow. In the future, we could detect this and emit the optimal instructions in the jit instead. But the code is already so fast that it might not be worth the effort.

The code is overall nearly 10x faster than the scalar version on my zen4 machine.

Before:
![image](https://github.com/user-attachments/assets/995a0b29-0a56-4511-a5c8-c8c2dc381b3b)
After:
![image](https://github.com/user-attachments/assets/db418171-a2eb-4674-afc1-20b4334e038a)

Before:
![image](https://github.com/user-attachments/assets/973a5c56-2c85-4f38-9831-ed71c03eddb6)
After:
![image](https://github.com/user-attachments/assets/b0c589a8-86d2-48bb-b527-c2d0e44b1ee9)

Whatcookie committed:
BufferUtils: Optimize upload_untouched_skip_restart with AVX-512 paths

- u16 path needs AVX-512-ICL because vpcompressw isn't included in skylake-x level AVX-512

- u32 path is untested as I couldn't find any games that hit it

Megamouse commented on rpcs3/Emu/RSX/Common/BufferUtils.cpp:
```diff
@@ -404,6 +419,153 @@ namespace
+	u32 min_index = _mm512_reduce_min_epu32(min);
+	u32 max_index = _mm512_reduce_max_epu32(max);
+
+return std::make_tuple(min_index, max_index, written);
```
```suggestion
	return std::make_tuple(min_index, max_index, written);
```

Megamouse reviewed (commented):
I think a lot of variables can be made const

AniLeo commented:
Tried NieR Replicant Mailbox and Fountain, Minecraft Menu and Tutorial, Diva F 2nd Menu, no performance difference on my side on any of these cases

kd-11 reviewed (commented):
All these paths have made this module which was meant to be a simple utils wrapper into an unmaintainable mess. The function names are also getting messy as intel adds more and more weird levels to the ISA.

Let's do this instead:
1. Move the different feature levels to separate files leaving the generic implementation here.
2. We expose a dispatch table for each feature level and pick the one to use at the start using a static lambda initializer. Some things to watch out for - Arm64 is actually sse4.2 compatible (including ssse3) when using sse2neon. Generic path is only required for validation as well as future architectures as a reference.
3. When intel releases avx9000 or whatever we just create a file for that featureset and don't keep adding to this file.

It's a lot more work but that's just how it is when you need to maintain a project.

Whatcookie commented:
> All these paths have made this module which was meant to be a simple utils wrapper into an unmaintainable mess. The function names are also getting messy as intel adds more and more weird levels to the ISA.
> 
> Let's do this instead:
> 
> 1. Move the different feature levels to separate files leaving the generic implementation here.
> 2. We expose a dispatch table for each feature level and pick the one to use at the start using a static lambda initializer. Some things to watch out for - Arm64 is actually sse4.2 compatible (including ssse3) when using sse2neon. Generic path is only required for validation as well as future architectures as a reference.
> 3. When intel releases avx9000 or whatever we just create a file for that featureset and don't keep adding to this file.
> 
> It's a lot more work but that's just how it is when you need to maintain a project.



> All these paths have made this module which was meant to be a simple utils wrapper into an unmaintainable mess. The function names are also getting messy as intel adds more and more weird levels to the ISA.
> 
> Let's do this instead:
> 
> 1. Move the different feature levels to separate files leaving the generic implementation here.
> 2. We expose a dispatch table for each feature level and pick the one to use at the start using a static lambda initializer. Some things to watch out for - Arm64 is actually sse4.2 compatible (including ssse3) when using sse2neon. Generic path is only required for validation as well as future architectures as a reference.
> 3. When intel releases avx9000 or whatever we just create a file for that featureset and don't keep adding to this file.
> 
> It's a lot more work but that's just how it is when you need to maintain a project.

I think I need to recover the old sse4.1 paths since neko removed them in favor of emitting x86 instructions directly in the jit, which won't work on arm

kd-11 commented:
The jit asm backend emits (or tries to) different instructions based on the hardware. It is supposed to be platform agnostic.

Whatcookie commented:
> The jit asm backend emits (or tries to) different instructions based on the hardware. It is supposed to be platform agnostic.

They're guarded by x86_64 ifdefs in this file, aren't they?

Megamouse commented:
@Whatcookie needs a rebase

Megamouse commented on rpcs3/Emu/RSX/Common/BufferUtils_avx512.h:
```diff
@@ -0,0 +1,158 @@
+#pragma once
```
Missing the vs files

Whatcookie replied:
Added

kd-11 commented on rpcs3/Emu/RSX/Common/BufferUtils_avx512.h:
```diff
@@ -0,0 +1,158 @@
+
+		max = _mm512_mask_max_epu32(max, mask, max, value);
+		min = _mm512_mask_min_epu32(min, mask, min, value);
+		const __m512i packed = _mm512_maskz_compress_epi32(mask, value_with_max_restart);
```
`const __m512i packed = _mm512_maskz_compress_epi32(mask, value);` achieves the same thing. You don't need the value_with_max_restart part anymore.
The mask set generated by the cmpneq_mask part should already generate a mask that omits the matches.
This applies to both this loop, the fallout one below and both u16 variants above.

Whatcookie replied:
oops

kd-11 commented on rpcs3/Emu/RSX/Common/BufferUtils_avx512.h:
```diff
@@ -0,0 +1,158 @@
+#define AVX512_ICL_FUNC __attribute__((__target__("avx512f,avx512bw,avx512dq,avx512cd,avx512vl,avx512bitalg,avx512ifma,avx512vbmi,avx512vbmi2,avx512vnni,avx512vpopcntdq")))
+#endif
+
+namespace
```
This is super-unsafe. But then again the whole header is very unsafe and needs to be included in a specific location in a specific file so maybe that's fine. If it's included in more than one TU - boom.
Just an observation for now really, I don't care either way.

Whatcookie commented:
Fixed the issue kd-11 raised. I don't have a compiler set up on my windows install right now, so I'll test after CI finishes.

Also added a comment about zen4 vcompress to memory being slow so this workaround doesn't get lost in the future
